cardinal/router/router.go

```diff
@@ -0,0 +1,238 @@
+package router
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/rotisserie/eris"
+	zerolog "github.com/rs/zerolog/log"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"net"
+	"pkg.world.dev/world-engine/cardinal/txpool"
+	shardtypes "pkg.world.dev/world-engine/evm/x/shard/types"
+	routerv1 "pkg.world.dev/world-engine/rift/router/v1"
+	shard "pkg.world.dev/world-engine/rift/shard/v2"
+	"pkg.world.dev/world-engine/sign"
+)
+
+const (
+	defaultPort = "9020"
+)
+
+//go:generate mockgen -source=router.go -package mocks -destination=mocks/router.go
+type Router interface {
+	Submit(
+		ctx context.Context,
+		processedTxs txpool.TxMap,
+		namespace string,
+		epoch,
+		unixTimestamp uint64,
+	) error
+
+	QueryTransactions(ctx context.Context, req *shardtypes.QueryTransactionsRequest) (
+		*shardtypes.QueryTransactionsResponse,
+		error,
+	)
+
+	Shutdown()
+	Run() error
+}
+
+var _ routerv1.MsgServer = (*routerImpl)(nil)
+var _ Router = (*routerImpl)(nil)
+
+type routerImpl struct {
+	routerv1.MsgServer
+
+	provider       Provider
+	ShardSequencer shard.TransactionHandlerClient
+	ShardQuerier   shardtypes.QueryClient
+
+	port     string
+	shutdown func()
+}
+
+func (r *routerImpl) Shutdown() {
+	if r.shutdown != nil {
+		r.shutdown()
+	}
+}
+
+func (r *routerImpl) Run() error {
+	server := grpc.NewServer()
+	routerv1.RegisterMsgServer(server, r)
+	listener, err := net.Listen("tcp", ":"+r.port)
+	if err != nil {
+		return eris.Wrapf(err, "error listening to port %s", r.port)
+	}
+	go func() {
+		err = eris.Wrap(server.Serve(listener), "error serving server")
+		if err != nil {
+			zerolog.Fatal().Err(err).Msg(eris.ToString(err, true))
+		}
+	}()
+	r.shutdown = server.GracefulStop
+	return nil
+}
+
+func New(sequencerAddr, baseShardQueryAddr string, provider Provider) (Router, error) {
```

Comment:
the constructor function should be the first function right below the type router struct {}. the methods should go after it in the order they are declared in the interface

Reply:
done